pkg/serviceresources/key.go

```diff
@@ -0,0 +1,183 @@
+package serviceresources
+
+import (
+	"context"
+
+	"go.uber.org/multierr"
+	"k8s.io/apimachinery/pkg/util/sets"
+
+	"github.com/seal-io/walrus/pkg/dao/model"
+	"github.com/seal-io/walrus/pkg/dao/model/connector"
+	"github.com/seal-io/walrus/pkg/dao/types"
+	"github.com/seal-io/walrus/pkg/dao/types/object"
+	"github.com/seal-io/walrus/pkg/operator"
+	optypes "github.com/seal-io/walrus/pkg/operator/types"
+	"github.com/seal-io/walrus/utils/gopool"
+	"github.com/seal-io/walrus/utils/log"
+)
+
+// SetKeys sets the keys of the resources for operations like log and exec.
+//
+// The given model.ServiceResource item must specify the following fields:
+// Shape, Mode, ID, DeployerType, Type, Name and ConnectorID.
+func SetKeys(
+	ctx context.Context,
+	logger log.Logger,
+	modelClient model.ClientSet,
+	candidates []*model.ServiceResource,
+	operators map[object.ID]optypes.Operator,
```

Comment:
Do we need this param? Seems like it is only used internally

Reply:
the graph of the environment.  https://github.com/seal-io/walrus/pull/1197/files#diff-1051d160d8e9e62077a18876598246711d6d5d7c236ac5e0d2d99f4ae94a91afR88-R96